### ISSUE TITLE
Better tables handling

### ISSUE THREAD

**Issue by fbordignon** [Slicer/Slicer#6128]
When adding (not so) big tables to Slicer, it takes a long time to show them. Example:

```python
import numpy as np
a = np.ones((10**6, 1))
tableNode = slicer.vtkMRMLTableNode()
updateTableFromArray(tableNode, a)
slicer.mrmlScene.AddNode(tableNode)
```

Go to the tables module and click the generated table to show it. Slicer takes more than 30 minutes to show this table, I don't know how long because after 30 min I've killed the process.

If a user saves this same table to a .csv on disk:

```python
import numpy as np
a = np.ones((10**6, 1))
np.savetxt("foo.csv", a, delimiter=",")
```
It saves it instantly on disk and loads in 3 seconds on ms excel, for example. The file generated is 25 MB, on memory should be 8 MB for float64.

I've tested on Windows 10 Slicer version 4.13.0 revision 30548 built 2022-01-24

**Comment by lassoan** [comment]
1-million rows is a very, very big for an editable table. That said, there is definitely lots of room for speed improvement.

The main performance bottleneck is the construction of qMRMLTableModel.

Quick workaround: See if you can easily activate some kind of batch update (with reset model, temporary signal blocking, etc.) when calling setMRMLTableNode. If there does not seem to be a quick way to do this with minimal change and extra complexities then give up on this approach and proceed to the proper solution.

Proper solution: Change base class of qMRMLTableModel from QStandardItemModel to QAbstractTableModel. This item model retrieves data items as needed instead of building a huge model upfront. This makes display and browsing of arbitrary sized tables immediate. It may even make the implementation simpler. You may use [ParaView's table model](https://github.com/Kitware/ParaView/blob/master/Qt/Core/pqSpreadSheetViewModel.cxx) as an example.

**Comment by fbordignon** [comment]
> 1-million rows is a very, very big for an editable table. That said, there is definitely lots of room for speed improvement.

I agree that a user will not edit a table adding 1 million rows by hand, but it is a small table for a result of some kind of data analysis or algorithm output. My point is that a table this size is fairly small for a modern computer to handle and should not effectively break Slicer upon request for visualization. The user may click it unintentionally or just want to browse the first rows to check the results, rename columns, convert column types, etc.

As always, thanks for the suggestions and comments. I will look into it.

**Comment by fbordignon** [comment]
Blocking signals on setMRMLTableNode lowers the time to show a table from 3 minutes to 8 seconds on a 100k lines single column table and from 12 minutes to 34 seconds on a 200k lines single column.
But it is still spending most of the time on https://github.com/Slicer/Slicer/blob/ae116729d4a69aa44927a3653bbbca36d5610993/Libs/MRML/Widgets/qMRMLTableModel.cxx#L386
If I comment this line, a table with a million rows is shown in 3 seconds.

**Comment by lassoan** [comment]
I'm afraid that this means that if we want near-real-time update then we must switch to QAbstractTableModel.

Response from Qt developers for questions about QStandardItemModel performance issue:

> QStandardItemModel is not meant to handle arbitrarily sized models, it's a convenience layer. If it's getting too slow, it's time to use a real QAbstractItemModel subclass.

https://bugreports.qt.io/browse/QTBUG-31581

It is somewhat frustrating because we got so close. Maybe you can spend a little time with a debugger and see if you can somehow block the unnecessary updates, but basically this is not a use case that this Qt class supports.

**Comment by fbordignon** [comment]
Right, I will try to fiddle a bit more with the setHeaderData.
If I understood correctly [this guy](https://www.qtcentre.org/threads/63038-rowCount()-is-calling-for-all-the-top-level-items?p=279789#post279789) said that it is possible to overwrite headerData() to return the desired data directly, without calling setHeaderData, or even overwriting it too. Maybe I will take a look into Qt implementation and simplify or prevent the calls that are taking too long.

**Comment by lassoan** [comment]
OK. I would just recommend to set a time limit for yourself with trying to fix the workaround, because the proper solution probably takes only a few hours to implement and the result is superior in all aspects.

**Comment by lassoan** [comment]
One complication with the workaround that it may only apply for the initial update (when you set the MRML node in the widget). If you change the table and update from MRML again then you may not avoid the very slow update: you cannot do a full update (you would lose selections, etc. in views) and you cannot block the signals (because then widgets would get out of sync). Using QAbstractTableModel should address all updates because Qt is smart enough to only ask for updates of the values that are visible in the views.

**Comment by lassoan** [comment]
Recently https://github.com/Kitware/VTK/blob/master/GUISupport/Qt/QVTKTableModelAdapter.h was added to VTK. We might be able to use it as is, or improve it to match what our table view can do now.
